[android] update listeners subscription lifecycle

arnaudvergnet opened the pull request:
Subscribe to listeners in onResume and unsubscribe in onPause.

Should hopefully fix #4729, but I was not able to reproduce it.


According to the [doc](https://developer.android.com/guide/fragments/lifecycle), these are the first/last events from the create/destroy lifecycle. So the callbacks should never get called when the views are not ready.

biodranik commented on android/src/app/organicmaps/widget/placepage/PlacePageView.java:
```diff
@@ -253,9 +250,17 @@ public void onAttach(@NonNull Context context)
+  public void onPause()
   {
-    super.onDestroyView();
+    super.onPause();
```
What other places?

@arnaudvergnet did you test it? @Jean-BaptisteC can you please help with tests by opening/closing/pausing/rotating the device with different placepages, with vulkan and opelgl modes?

arnaudvergnet replied:
> What other places?

In each child fragment of the place page (wiki, opening hours, bookmark...). We subscribe to the map object viewmodel so we need to use the same lifecycle methods, I'll update the PR.

arnaudvergnet commented:
I've updated the pr to call onResume/onPause in all place page child fragments, and also moved the call to detachCountry to onPause (see associated comment) because I think it is safer to call here.

biodranik commented on android/src/app/organicmaps/widget/placepage/PlacePageView.java:
```diff
@@ -253,14 +250,34 @@ public void onAttach(@NonNull Context context)
+    // It is safer to call this here than in onPause as the app could go from onPause to
+    // onResume without killing the fragment.
+    // In this case we would not want to detach the country.
+    detachCountry();
```
Storage callbacks are using a lot of unnecessary CPU. Why it can't be detached in onPause and attached again in onResume?

arnaudvergnet replied:
onPause may be called while the app is still visible to the user, while onStop is only called when the activity is fully covered by another (after onPause). We should never call CPU intensive functions inside onPause because it could slow down the transition animation (see https://developer.android.com/reference/android/app/Activity.html#onPause()).

So here the country is only attached when the current map object changes, and detached when the activity/fragment stops.

biodranik reviewed (approved):
CC @rtsisyk

arnaudvergnet commented:
Updated the PR because I forgot the phone fragment. I'll move all place page fragments in their own package in another PR so we don't forget about one again

biodranik reviewed (approved):
@rtsisyk let's merge it?

Jean-BaptisteC reviewed (approved):
✅ Pixel 6 - Android 13
